src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -203,15 +196,83 @@ void NnetDecodableBase::GetCurrentIvector(int32 output_t_start,
 }
 
+void NnetDecodableBase::GetIvectorsForFrames(int32 output_t_start,
+                                             int32 num_output_frames,
+                                             Matrix<BaseFloat> *ivectors) {
+  const int32 left_context = nnet_left_context_ + opts_.extra_left_context,
+        right_context = nnet_right_context_ + opts_.extra_right_context;
+  // frame_to_search is the frame that we want to get the most recent iVector
+  // for.  We choose a point near the middle of the current window, the concept
+  // being that this is the fairest comparison to nnet2.   Obviously we could do
+  // better by always taking the last frame's iVector, but decoding with
+  // 'online' ivectors is only really a mechanism to simulate online operation.
+  if (ivector_interval_ == 0) { // single ivector case
+    ivectors->Resize(1, online_ivector_feats_->NumCols(), kUndefined);
+    int32 frame_to_search = output_t_start + num_output_frames / 2;
+    SubVector<BaseFloat> sub_ivector = ivectors->Row(0);
+    GetCurrentIvector(frame_to_search, &sub_ivector);
+  } else { // multiple ivectors case
+    // num_ivectors_left is the num of ivectors for frames whose "t" index < 0.
+    // num_ivectors_right is the num of ivectors for frames whose "t" index >= 0
+    // They are computed according to how Round descriptor works, which
+    // basically returns floor(t / <t-modulus>) * <t-modulus>.
+    // (The assumption is that the t index of the first outptut frame is 0)
+    int32 num_ivectors_left = std::ceil(1.0 * left_context / ivector_interval_);
+    int32 num_ivectors_right = (num_output_frames + right_context - 1) /
+                               ivector_interval_ + 1;
+    ivectors->Resize(num_ivectors_left + num_ivectors_right,
+                    online_ivector_feats_->NumCols(), kUndefined);
+    int32 frame_to_search = output_t_start - 1 -
+                            std::min(ivector_interval_, left_context) / 2;
+    for (int32 n = 0; n < num_ivectors_left; n++) {
+      int32 row = num_ivectors_left - 1 - n;
+      SubVector<BaseFloat> sub_ivector = ivectors->Row(row);
+      GetCurrentIvector(frame_to_search, &sub_ivector);
+      frame_to_search -= ivector_interval_;
+    }
+    frame_to_search = output_t_start +
+                      std::min(ivector_interval_, num_output_frames) / 2;
+    for (int32 n = 0; n < num_ivectors_right; n++) {
+      int32 row = num_ivectors_left + n;
+      SubVector<BaseFloat> sub_ivector = ivectors->Row(row);
+      GetCurrentIvector(frame_to_search, &sub_ivector);
+      frame_to_search += ivector_interval_;
+    }
+  }
+}
+
+void NnetDecodableBase::GenerateIndexesForIvectors(
+    int32 num_ivectors, std::vector<Index> *indexes) {
+  if (ivector_interval_ == 0)
+    indexes->push_back(Index(0, 0, 0));
+  else {
+    int32 left_context = nnet_left_context_ + opts_.extra_left_context;
+    int32 num_ivectors_left = std::ceil(1.0 * left_context / ivector_interval_);
```

Comment:
There is no need for this variable at all-- for some reason Yiming seems to
think the left and right context need to be handle separately, which is not
the case.  The same issue was in the first draft of this code.

On Thu, Mar 10, 2016 at 8:55 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-am-decodable-simple.cc
> <https://github.com/kaldi-asr/kaldi/pull/468#discussion_r55780395>:
>
> > +    for (int32 n = 0; n < num_ivectors_right; n++) {
> > +      int32 row = num_ivectors_left + n;
> > +      SubVector<BaseFloat> sub_ivector = ivectors->Row(row);
> > +      GetCurrentIvector(frame_to_search, &sub_ivector);
> > +      frame_to_search += ivector_interval_;
> > +    }
> > +  }
> > +}
> > +
> > +void NnetDecodableBase::GenerateIndexesForIvectors(
> > +    int32 num_ivectors, std::vector<Index> *indexes) {
> > +  if (ivector_interval_ == 0)
> > +    indexes->push_back(Index(0, 0, 0));
> > +  else {
> > +    int32 left_context = nnet_left_context_ + opts_.extra_left_context;
> > +    int32 num_ivectors_left = std::ceil(1.0 * left_context / ivector_interval_);
>
> while reading this code I was confused by the name num_ivector_left,
> better use num_ivectors_in_left_context
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/468/files#r55780395>.
>